### PR TITLE
feat(ts): provide type checking through `fork-ts-checker-webpack-plugin`

### PR DESCRIPTION
**Pull request by kevinmarrec**
## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

This PR improves Nuxt TS support by adding a build option `useForkTschecker` that enable type checking thanks to [fork-ts-checker-webpack-plugin](https://www.npmjs.com/package/fork-ts-checker-webpack-plugin).

Type checking makes you able to know if there is TS related errors within your code, which means that without it you may have Nuxt Webpack compilation succeed, but the errors only reported at runtime and not in an elegant way.

The type checker will check your code each time you make changes. In other terms, it's trigerred after each new Webpack compilation.

### Usage

```js
// nuxt.config.js
export default {
  build: {
    useForkTsChecker: true
  }
}
```

`useForkTsChecker` can either be a boolean or an object. If it's an object, it will override the [plugin options](https://www.npmjs.com/package/fork-ts-checker-webpack-plugin#options)

```js
// nuxt.config.js
export default {
  build: {
    useForkTsChecker: {
      silent: true,
      tslint: false
    }
  }
}
```

### Example of an error reported by the type checker

screenshot_1

### Project requirements to use this feature

The `fork-ts-checker-webpack-plugin` is used/enabled only if the user project meets these requirements :
- `useForkTsChecker` option is enabled (true) in the Nuxt configuration file
- `ts-loader` is configured with the `transpileOnly` mode (to not have double type checkers)
- `fork-ts-checker-webpack-plugin` (and also `tslint`) devDependencies are installed

If the two first rules are fulfilled but not the latter, a warning will let you know that your project is missing the dependencies required :

screenshot_10

### Documentation

Documentation related to this PR will come later as a whole TS support documentation.

## Checklist:

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [x] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests are passing.

### CODE REVIEW

File: packages/common/src/utils.js
Comment (pi0) [inline packages/common/src/utils.js]: We have this util in nuxt.resolver

File: packages/webpack/src/config/base.js
Comment (pi0) [inline packages/webpack/src/config/base.js]: Suggest adding a comment above this line

File: packages/webpack/src/config/base.js
Comment (pi0) [inline packages/webpack/src/config/base.js]: Damn 😮 It natively supports vue 👍

File: packages/webpack/src/config/base.js
Comment (pi0) [inline packages/webpack/src/config/base.js]: nuxt-config and other configs are usually resolved from rootDir. OR it should be better to check both, like postcss config

File: packages/webpack/src/config/base.js
Comment (pi0) [inline packages/webpack/src/config/base.js]: use consola (and no Eslint will be required)

File: test/fixtures/typescript/nuxt.config.js
Comment (pi0) [inline test/fixtures/typescript/nuxt.config.js]: Approve as a non breaking change as 2.4 is not released.

**Review by pi0 — CHANGES_REQUESTED** [review]
PR description is 💯

**Comment by kevinmarrec** [comment]
@pi0 Thanks for your review ! I'm going to make the needed changes :)

File: packages/webpack/src/config/base.js
Comment (kevinmarrec) [reply packages/webpack/src/config/base.js]: 😁

File: packages/common/src/utils.js
Comment (kevinmarrec) [reply packages/common/src/utils.js]: @pi0 `requireModule` from `nuxt.resolver` throws an error if the package is not found.
I wanted to find a way to continue the process with a warning instead of making the build fail. What do you suggest then ?

File: packages/webpack/src/config/base.js
Comment (kevinmarrec) [reply packages/webpack/src/config/base.js]: Alright ! Will use `rootDir` instead !

File: test/fixtures/typescript/nuxt.config.js
Comment (kevinmarrec) [reply test/fixtures/typescript/nuxt.config.js]: There is no changes to do around here right ?

File: packages/common/src/utils.js
Comment (clarkdo) [reply packages/common/src/utils.js]: use `resolveModule` in `resolver`

File: test/fixtures/typescript/nuxt.config.js
Comment (pi0) [reply test/fixtures/typescript/nuxt.config.js]: Yeah. Commented to confirm this change. 🙈

**Review by clarkdo — DISMISSED** [review]
LGTM except change `tryRequire` to `resolver.resolveModule`

**Comment by clarkdo** [comment]
Kind of, and you can use the return value of resolveModule if it exists, then the require won’t resolve all the paths again.

**Comment by kevinmarrec** [comment]
Wait before merging, I need to test something :P

**Comment by kevinmarrec** [comment]
@pi0 @clarkdo Alright, need final review & LGTM :)

**Review by hartmut-co-uk — DISMISSED** [review]
LGTM

**Review by pi0 — DISMISSED** [review]

**Review by clarkdo — DISMISSED** [review]

File: packages/webpack/src/config/base.js
Comment (clarkdo) [inline packages/webpack/src/config/base.js]: If the plugin in only for `client`, this part can be moved to `client.js` 😄

**Comment by kevinmarrec** [comment]
@clarkdo @pi0 I moved the plugin in `client.js` as suggested 🙂
(I forced push to revert a build-failing commit while keeping Git history clean, I tried to remove `tslint` devDependency but was not a good idea RIP 😅)

**Review by pi0 — APPROVED** [review]

**Comment by hartmut-co-uk** [comment]
🎊

**Comment by husayt** [comment]
@kevinmarrec one more thing, with eslint it was recommended to also add `eslint-loader` in `nuxt-config.js`, is there any value of adding `tslint-loader` i.e.

```
      if (isDev && isClient) {
        config.module.rules.push({
          enforce: "pre",
          test: /\.(ts|js|vue)$/,
          loader: "tslint-loader",
          exclude: /(node_modules)/
        })
      }

```

**Comment by kevinmarrec** [comment]
@husayt This behavior is likely what `ForkTsChecker` does if you configure the `tslint` option of it :)

So no `tslint-loader` needed, `ForkTsChecker` will check if your files have lint errors whenever you change them 😉

**Pull request merged**
